src/sentry/api/endpoints/artifact_lookup.py

```diff
@@ -0,0 +1,213 @@
+                    if not url_in_any_manifest(url):
+                        missing_urls.append(url)
+
+                # Possibly use the algorithm sketched up here:
+                # https://github.com/getsentry/sentry/pull/45697#issuecomment-1466389132
+                # That would narrow down our set of bundles to the minimum set that covers
+                # the file names we are querying for, and also leave us with the remaining
+                # set of file names that are not covered by any bundle, to look up below
+
+        # Second, look for a legacy `ReleaseFile` (or whatever) if an individual
+        # file exists matching the release/dist/url we are looking for.
+
+        # TODO: also query for and return individual artifacts
+        individual_files = set()
+
+        # Then: Construct our response
+        url_constructor = UrlConstructor(request)
+
+        found_artifacts = []
+        for file_id in bundle_file_ids:
+            found_artifacts.append(
+                {
+                    "type": "bundle",
+                    "url": url_constructor.url_for_file_id(file_id),
+                }
+            )
+
+        for file in individual_files:
+            found_artifacts.append(
+                {
+                    "type": "file",
+                    "url": url_constructor.url_for_file_id(file.id),
+                    # The `name` is the url/abs_path of the file,
+                    # as in: `"~/path/to/file.min.js"`.
+                    "abs_path": file.name,
+                    # These headers should ideally include the `Sourcemap` reference
+                    "headers": file.headers,
+                }
+            )
+
+        # NOTE: We do not paginate this response, as we have very tight limits
+        # on all the individual queries.
+        return Response(serialize(found_artifacts, request.user))
+
+
+def url_exists_in_manifest(manifest: dict, url: str) -> bool:
+    """
+    Looks through all the files in the `ArtifactBundle` manifest and see if the
+    `url` matches any of the files.
+    """
+    try:
+        files = manifest.get("files", dict())
+        for file in files.values():
+            if url in file.get("url", ""):
+                return True
+    except Exception:
+        return False
+
+
+class UrlConstructor:
+    def __init__(self, request: Request):
+        # TODO: is there a better way to construct a url to this same route?
+        self.base_url = request.build_absolute_uri(request.path)
+
+    def url_for_file_id(self, file_id: int) -> str:
+        # NOTE: Returning a self-route that requires authentication (via Bearer token)
+        # is not really forward compatible with a pre-signed URL that does not
+        # require any authentication or headers whatsoever.
+        # This also requires a workaround in Symbolicator, as its generic http
+        # downloader blocks "internal" IPs, whereas the internal Sentry downloader
+        # is explicitly exempt.
```

Comment:
This should be fairly easy to work around in symbolicator. We can just check `starts_with` and use one downloader with token or the other…